docs/design/extending-api.md

```diff
@@ -0,0 +1,227 @@
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/design/extending-api.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Adding custom resources to the Kubernetes API server
+
+This document describes the design for implementing the storage of custom API types in the Kubernetes API Server.
+
+
+## Resource Model
+
+### The ThirdPartyResource
+
+The `ThirdPartyResource` resource describes the multiple versions of a custom resource that the user wants to add
+to the Kubernetes API.  `ThirdPartyResource` is a non-namespaced resource, attempting to place it in a resource
+will return an error.
+
+Each `ThirdPartyResource` resource has the following:
+   * Standard Kubernetes object metadata.
+   * ResourceKind - The kind of the resources described by this third party resource.
+   * Description - A free text description of the resource.
+   * APIGroup - An API group that this resource should be placed into.
+   * Versions - One or more `Version` objects.
+
+### The `Version` Object
+
+The `Version` object describes a single concrete version of a custom resource.  The `Version` object currently
+only specifies:
+   * The `Name` of the version.
+   * The `APIGroup` this version should belong to.
+
+## Expectations about third party objects
+
+Every object that is added to a third-party Kubernetes object store is expected to contain Kubernetes
+compatible [object metadata](../devel/api-conventions.md#metadata).  This requirement enables the
+Kubernetes API server to provide the following features:
+   * Filtering lists of objects via LabelQueries
+   * `resourceVersion`-based optimistic concurrency via compare-and-swap
+   * Versioned storage
+   * Event recording
+   * Integration with basic `kubectl` command line tooling.
+   * Watch for resource changes.
+
+The `Kind` for an instance of a third-party object (e.g. CronTab) below is expected to be
+programnatically convertible to the name of the resource using
+the following conversion.  Kinds are expected to be of the form:
+
+```
+<domain-name>/<CamelCaseKind>
+```
+
+`domain-name` is expected to be a fully qualified domain name.
+
+'CamelCaseKind' is the specific type name.
+
+To convert this into the name for the `ThirdPartyResource` resource instance,
+the `<domain-name>` and `/` are copied verbatim, the `CamelCaseKind` is
+then converted
+using '-' instead of capitalization ('camel-case'), with the first character being assumed to be
+capitalized.  In pseudo code:
+
+```go
+var result string
+for ix := range kindName {
+  if isCapital(kindName[ix]) {
+    result = append(result, '-')
+  }
+  result = append(result, toLowerCase(kindName[ix])
+}
+```
+
+As a concrete example, the resource named `example.com/camel-case-kind` defines resources of Kind `example.com/CamelCaseKind`.
```

Comment:
The resource can't be named `example.com/camel-case-kind`. We don't allow `/` in names. 

More importantly, the domain (`example.com`) logically should be part of the APIGroup rather than the Kind. (It's similar to a "provider" concept that we've discussed in #7111.)

cc @nikhiljindal @caesarxuchao @uluyol